### PR TITLE
Issue #14019: Kill this.checks = null mutation on SuppressionXpathSingleFilter

### PR DESCRIPTION
**Pull request by kkoutsilis**
Part of #14019 
Removed suppression
https://github.com/checkstyle/checkstyle/blob/954a9f121861cd71ddd8838fa2233384e62ded12/config/pitest-suppressions/pitest-filters-suppressions.xml#L165-L172

Followed https://github.com/checkstyle/checkstyle/pull/13794 as they are similar

## Steps followed to kill mutation 
This steps are better described [here](https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports),
- Removed suppression from `pitest-filters-suppressions.xml`
- Listed pitest profiles with `groovy .ci/pitest-survival-check-xml.groovy --list` using java 11 and groovy 2.4.5
- Picked up `pitest-filters` as active profile
- Generated report locally using `mvn -e --no-transfer-progress -P"$PITEST_PROFILE" clean test-compile org.pitest:pitest-maven:mutationCoverage`  
  <details>
  <summary> Screenshots </summary>
  
  Screenshot from 2024-10-15 20-49-37 
  
  Screenshot from 2024-10-15 19-14-54
  </details>

- Analyzed report localy using `groovy .ci/pitest-survival-check-xml.groovy "$PITEST_PROFILE"`. Response:
	```shell
	Source File: "SuppressionXpathSingleFilter.java"
	Class: "com.puppycrawl.tools.checkstyle.filters.SuppressionXpathSingleFilter"
	Method: "setChecks"
	Line Contents: "this.checks = null;"
	Mutator: "org.pitest.mutationtest.engine.gregor.mutators.experimental.MemberVariableMutator"
	Description: "Removed assignment to member variable checks"
	Line Number: 141
	```
- Added extra test case
- Generated report again
    <details>
    <summary> Screenshots </summary>
    
    
    Screenshot from 2024-10-15 20-37-32
    Screenshot from 2024-10-15 20-35-18
</details>

- Report analysis `No new surviving mutation(s) found.`

### CODE REVIEW

**Comment by kkoutsilis** [comment]
Check PR Description failed because I did not [put a reference to the issue in the description ](https://github.com/checkstyle/checkstyle/actions/runs/11353537553/job/31578773462?pr=15777#step:3:11). Description is updated to reference issue.

**Review by romani — CHANGES_REQUESTED** [review]
Our API name for filter method is supper confusing.
Very hard to understand code on either true or false should be returned. Very unfortunate, but too late to improve , it might be big breaking change.

https://github.com/checkstyle/checkstyle/blob/954a9f121861cd71ddd8838fa2233384e62ded12/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L245-L246

https://github.com/checkstyle/checkstyle/blob/954a9f121861cd71ddd8838fa2233384e62ded12/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilter.java#L186-L187


Items:

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java [hunk 1]
```diff
@@ -446,9 +446,50 @@ public void testUpdateFilterSettingsInRunTime() throws Exception {
...
+
+        assertWithMessage("match is expected as 'checks' is set")
                 .that(filter.accept(ev))
                 .isFalse();
+
+        // reset checks to default value of filter
+        filter.setChecks(null);
+        filter.finishLocalSetup();
+
+        assertWithMessage("no match is expected as 'checks' is defaulted")
```
[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java:490]: https://github.com/checkstyle/checkstyle/blob/954a9f121861cd71ddd8838fa2233384e62ded12/src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElement.java#L125-L126

[the remark this answers is not in the document]
Comment (kkoutsilis) [reply src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java]: Updated

**Review by romani — APPROVED** [review]
Ok to merge.

Thanks a lot for very clear and clean PR

**Comment by kkoutsilis** [comment]
@romani any preference on the next mutation to pick up from the list? Or any other issue I can help with?

**Comment by romani** [comment]
@kkoutsilis , thanks a lot for help.
please try https://github.com/checkstyle/checkstyle/issues/13999 first item for JavadocMethodCheck.
it should result in update for Inputs files, and may be we will need testing diff githib workflow thatt ease testing on big real projects.
Beware this Check is very old, it is still based on regexp parsing of javadoc, we should try to avoid being involved in massive refactoring of it in scope of that issue.

**Review by rdiachenko — APPROVED** [review]
lgtm

**Pull request merged**